File: src/python/analisis_fft.py
import numpy as np
import matplotlib.pyplot as plt

# =========================
# Parámetros generales
# =========================
fs = 1000            # Frecuencia de muestreo [Hz]
T = 1.0              # Duración [s]
N = int(fs * T)      # Muestras totales
t = np.arange(N) / fs

def fft_mag_phase(x, fs):
    N = len(x)
    X = np.fft.fft(x)
    f = np.fft.fftfreq(N, d=1/fs)
    # Ordenar para espectro centrado (opcional)
    X_shift = np.fft.fftshift(X)
    f_shift = np.fft.fftshift(f)
    mag = np.abs(X_shift)
    phase = np.angle(X_shift)
    return f_shift, mag, phase

def plot_time_and_freq(t, x, f, mag, phase, title_base, save_prefix):
    plt.figure(figsize=(12, 8))
    plt.subplot(2,2,1)
    plt.plot(t, x, lw=1.5)
    plt.title(f"{title_base} - Tiempo")
    plt.xlabel("Tiempo [s]"); plt.ylabel("Amplitud"); plt.grid(True)

    plt.subplot(2,2,2)
    plt.plot(f, mag, lw=1.5)
    plt.title(f"{title_base} - Magnitud")
    plt.xlabel("Frecuencia [Hz]"); plt.ylabel("|X(f)|"); plt.grid(True)

    plt.subplot(2,2,4)
    plt.plot(f, phase, lw=1.0)
    plt.title(f"{title_base} - Fase")
    plt.xlabel("Frecuencia [Hz]"); plt.ylabel("∠X(f) [rad]"); plt.grid(True)

    plt.tight_layout()
    plt.savefig(f"figures/{save_prefix}.png", dpi=150)
    plt.show()

# =========================
# 1) Señal senoidal
# =========================
f0 = 50
A = 1.0
x_sine = A * np.sin(2*np.pi*f0*t)
f_sine, mag_sine, phase_sine = fft_mag_phase(x_sine, fs)
plot_time_and_freq(t, x_sine, f_sine, mag_sine, phase_sine, "Senoidal 50 Hz", "senoidal_50Hz")

# =========================
# 2) Escalón (u[n])
# =========================
# En tiempo continuo, el escalón no es absolutamente integrable; discretamente lo truncamos.
x_step = np.ones_like(t)
f_step, mag_step, phase_step = fft_mag_phase(x_step, fs)
plot_time_and_freq(t, x_step, f_step, mag_step, phase_step, "Escalón truncado", "escalon_truncado")

# =========================
# 3) Pulso rectangular
# =========================
# Pulso de ancho 0.2 s centrado al inicio (0 <= t < 0.2)
width = 0.2
x_rect = np.where((t >= 0.0) & (t < width), 1.0, 0.0)
f_rect, mag_rect, phase_rect = fft_mag_phase(x_rect, fs)
plot_time_and_freq(t, x_rect, f_rect, mag_rect, phase_rect, "Pulso rectangular (0.2 s)", "pulso_rectangular")

# =========================
# Propiedad de linealidad: a*x1 + b*x2 -> a*X1 + b*X2
# =========================
a, b = 0.7, 0.3
x_lin = a*x_sine + b*x_rect
f_lin, mag_lin, phase_lin = fft_mag_phase(x_lin, fs)
plot_time_and_freq(t, x_lin, f_lin, mag_lin, phase_lin, "Linealidad: 0.7*seno + 0.3*pulso", "linealidad")

# =========================
# Propiedad de desplazamiento en el tiempo: x(t - t0) -> X(f) * e^{-j2π f t0}
# =========================
t0 = 0.1
# Desplazamiento discreto: desplazar índices
shift_samples = int(t0 * fs)
x_rect_shift = np.roll(x_rect, shift_samples)
f_shifted, mag_shifted, phase_shifted = fft_mag_phase(x_rect_shift, fs)
plot_time_and_freq(t, x_rect_shift, f_shifted, mag_shifted, phase_shifted, "Pulso desplazado 0.1 s", "pulso_desplazado")

# =========================
# Escalamiento en el tiempo (equivale a escalamiento en frecuencia)
# x(a t) <-> (1/|a|) X(f / a)
# =========================
# A nivel discreto, simular 'comprimir' la señal duplicando frecuencia efectiva
f0_2 = 100
x_sine_fast = A * np.sin(2*np.pi*f0_2*t)
f_fast, mag_fast, phase_fast = fft_mag_phase(x_sine_fast, fs)
plot_time_and_freq(t, x_sine_fast, f_fast, mag_fast, phase_fast, "Senoidal 100 Hz (escalado en tiempo)", "senoidal_100Hz")

# =========================
# Comparación de dominio de frecuencia: ventana y padding (mejora visual)
# =========================
# Ventana Hamming para el seno de 50 Hz
w = np.hamming(N)
x_sine_win = x_sine * w
f_win, mag_win, phase_win = fft_mag_phase(x_sine_win, fs)
plot_time_and_freq(t, x_sine_win, f_win, mag_win, phase_win, "Senoidal 50 Hz con ventana Hamming", "senoidal_ventana")

# Zero-padding para mayor resolución en frecuencia
pad_factor = 4
X_pad = np.fft.fft(x_sine, n=N*pad_factor)
f_pad = np.fft.fftfreq(N*pad_factor, d=1/fs)
X_pad_shift = np.fft.fftshift(X_pad)
f_pad_shift = np.fft.fftshift(f_pad)
mag_pad = np.abs(X_pad_shift)
phase_pad = np.angle(X_pad_shift)

plt.figure(figsize=(10,6))
plt.plot(f_sine, mag_sine, label="Sin padding", alpha=0.7)
plt.plot(f_pad_shift, mag_pad, label=f"Padding x{pad_factor}", alpha=0.7)
plt.title("Comparación de magnitud (seno 50 Hz) con/ sin padding")
plt.xlabel("Frecuencia [Hz]"); plt.ylabel("|X(f)|"); plt.grid(True); plt.legend()
plt.tight_layout()
plt.savefig("figures/senoidal_padding.png", dpi=150)
plt.show()

File: src/matlab/analisis_fft.m
% Parámetros
fs = 1000; T = 1.0; N = fs*T; t = (0:N-1)/fs;

%% 
% Funciones utilitarias
fft_mag_phase = @(x) deal(fftshift((-fs/2):(fs/N):(fs/2 - fs/N)), ...
    abs(fftshift(fft(x))), angle(fftshift(fft(x))));

function plot_time_freq(t,x,f,mag,phase,title_base,save_prefix)
    figure('Position',[100 100 1000 600]);
    subplot(2,2,1); plot(t,x,'LineWidth',1.5); title([title_base ' - Tiempo']); xlabel('Tiempo [s]'); ylabel('Amplitud'); grid on;
    subplot(2,2,2); plot(f,mag,'LineWidth',1.5); title([title_base ' - Magnitud']); xlabel('Frecuencia [Hz]'); ylabel('|X(f)|'); grid on;
    subplot(2,2,4); plot(f,phase,'LineWidth',1.0); title([title_base ' - Fase']); xlabel('Frecuencia [Hz]'); ylabel('∠X(f) [rad]'); grid on;
    set(gcf,'Color','w'); if ~exist('figures','dir'), mkdir('figures'); end
    saveas(gcf, fullfile('figures',[save_prefix '.png']));
end

% Señales
f0 = 50; A = 1.0; x_sine = A*sin(2*pi*f0*t);
[f_sine, mag_sine, phase_sine] = fft_mag_phase(x_sine); plot_time_freq(t,x_sine,f_sine,mag_sine,phase_sine,'Senoidal 50 Hz','senoidal_50Hz_matlab');

x_step = ones(size(t));
[f_step, mag_step, phase_step] = fft_mag_phase(x_step); plot_time_freq(t,x_step,f_step,mag_step,phase_step,'Escalón truncado','escalon_truncado_matlab');

width = 0.2; x_rect = double((t >= 0.0) & (t < width));
[f_rect, mag_rect, phase_rect] = fft_mag_phase(x_rect); plot_time_freq(t,x_rect,f_rect,mag_rect,phase_rect,'Pulso rectangular (0.2 s)','pulso_rectangular_matlab');

% Propiedades
a=0.7; b=0.3; x_lin = a*x_sine + b*x_rect;
[f_lin, mag_lin, phase_lin] = fft_mag_phase(x_lin); plot_time_freq(t,x_lin,f_lin,mag_lin,phase_lin,'Linealidad: 0.7*seno + 0.3*pulso','linealidad_matlab');

t0 = 0.1; shift_samples = round(t0*fs); x_rect_shift = circshift(x_rect, [0 shift_samples]);
[f_shift, mag_shift, phase_shift] = fft_mag_phase(x_rect_shift); plot_time_freq(t,x_rect_shift,f_shift,mag_shift,phase_shift,'Pulso desplazado 0.1 s','pulso_desplazado_matlab');

f0_2 = 100; x_sine_fast = A*sin(2*pi*f0_2*t);
[f_fast, mag_fast, phase_fast] = fft_mag_phase(x_sine_fast); plot_time_freq(t,x_sine_fast,f_fast,mag_fast,phase_fast,'Senoidal 100 Hz (escalado en tiempo)','senoidal_100Hz_matlab');

% Ventanas y padding
w_names = {'rectwin','hamming','hann','blackman'};
figure('Position',[100 100 900 500]); hold on;
for k=1:numel(w_names)
    w = window(str2func(w_names{k}), N);
    Xw = fftshift(fft(x_sine(:).*w(:)));
    f = fftshift(((0:N-1)/N)*fs - fs/2);
    plot(f, abs(Xw), 'DisplayName', w_names{k});
end
title('Comparación de ventanas (seno 50 Hz)'); xlabel('Frecuencia [Hz]'); ylabel('|X(f)|'); grid on; legend; set(gcf,'Color','w');
saveas(gcf, fullfile('figures','ventanas_comparacion_matlab.png'));

pad_factor = 4;
X_pad = fft(x_sine, N*pad_factor);
f_pad = fftshift(((0:N*pad_factor-1)/(N*pad_factor))*fs - fs/2);
figure('Position',[100 100 900 500]); 
plot(f_sine, mag_sine, 'DisplayName','Sin padding'); hold on;
plot(f_pad, abs(fftshift(X_pad)), 'DisplayName',['Padding x' num2str(pad_factor)]);
title('Magnitud con y sin padding (seno 50 Hz)'); xlabel('Frecuencia [Hz]'); ylabel('|X(f)|'); grid on; legend; set(gcf,'Color','w');
saveas(gcf, fullfile('figures','senoidal_padding_matlab.png'));